classes/quiz_settings.php

```diff
@@ -147,6 +152,193 @@ protected static function define_properties() {
+        $map = $this->get_bool_seb_setting_map();
+        foreach ($settings as $setting => $value) {
+            if (isset($map[$setting])) {
+                $enabled = $value === 1 ? true : false;
+                $plist->add_element_to_root($map[$setting], new CFBoolean($enabled));
+            }
+        }
+    }
+
+    /**
+     * Turn hashed quit password and quit link into PList strings and add to config PList.
+     *
+     * @param property_list $plist The config PList.
+     * @param \stdClass $settings The quiz settings.
+     */
+    private function process_quit_settings(property_list $plist, \stdClass $settings) {
+        if (!empty($settings->quitpassword) && is_string($settings->quitpassword)) {
+            // Hash quit password.
+            $hashedpassword = hash('SHA256', $settings->quitpassword);
+            $plist->add_element_to_root('hashedQuitPassword', new CFString($hashedpassword));
+        }
+
+        if (!empty($settings->linkquitseb) && is_string($settings->linkquitseb)) {
+            $plist->add_element_to_root('quitURL', new CFString($settings->linkquitseb));
+        }
+    }
+
+    /**
+     * Turn return separated strings for URL filters into a PList array and add to config PList.
+     *
+     * @param property_list $plist The config PList.
+     * @param \stdClass $settings The quiz settings.
+     */
+    private function process_url_filters(property_list $plist, \stdClass $settings) {
+        // Create rules to each expression provided and add to config.
+        $urlfilterrules = [];
+        // Get all rules separated by newlines and remove empty rules.
+        $expallowed = array_filter(explode(PHP_EOL, $settings->expressionsallowed));
+        $expblocked = array_filter(explode(PHP_EOL, $settings->expressionsblocked));
+        $regallowed = array_filter(explode(PHP_EOL, $settings->regexallowed));
+        $regblocked = array_filter(explode(PHP_EOL, $settings->regexblocked));
+        foreach ($expallowed as $exp) {
+            $urlfilterrules[] = $this->create_filter_rule(true, false, $exp);
+        }
+        foreach ($expblocked as $exp) {
+            $urlfilterrules[] = $this->create_filter_rule(false, false, $exp);
+        }
+        foreach ($regallowed as $exp) {
+            $urlfilterrules[] = $this->create_filter_rule(true, true, $exp);
+        }
+        foreach ($regblocked as $exp) {
+            $urlfilterrules[] = $this->create_filter_rule(false, true, $exp);
+        }
+        $plist->add_element_to_root('URLFilterRules', new CFArray($urlfilterrules));
+    }
+
+    /**
+     * Create a CFDictionary represeting a URL filter rule.
+     *
+     * @param bool $allowed Allowed or blocked.
+     * @param bool $regex Regex or simple.
+     * @param string $exp The expression to filter with.
+     * @return CFDictionary A PList dictionary.
+     */
+    private function create_filter_rule(bool $allowed, bool $regex, string $exp) : CFDictionary {
```

Comment:
I'd suggest to change the order of the vars:  to  create_filter_rule(string $exp, bool $allowed, bool $regex)
First we tell what the rule is, then we tell what to do with the rule and then we tell how to read the rule.

Reply:
Done